Feature/PO-6 - Model Block, Transaction and Blockchain

javierdiaz72 opened the pull request:
The intention of this PR is to provide an implementation of transactions, blocks and blockchains, along with a function to evolve a given genesis block into a new genesis block by applying the transactions contained in the blocks of a given blockchain.

Commit:
Initial modeling of transactions

Commit:
Implemented initial modeling of transactions.

Commit:
Merge branch 'feature/PO4' of github.com:atixlabs/praos_test into feature/PO-6-Model-Block-Transaction-Blockchain

Commit:
Initial modeling of hashing functions.

Commit:
Merge branch 'feature/PO4' of github.com:atixlabs/praos_test into feature/PO-6-Model-Block-Transaction-Blockchain

Commit:
Initial implementation

Commit:
Merge branch 'develop' of github.com:atixlabs/praos_test into feature/PO-6-Model-Block-Transaction-Blockchain

ntallar commented on src/main/scala/io/iohk/praos/crypto/package.scala:
```diff
@@ -59,5 +59,44 @@ package object crypto {
+  /**
+    * A calculator of hash values.
+    */
+  trait Hasher {
```
I'm not sure if this trait is needed, are we expecting to extend it later?

javierdiaz72 replied:
If we want to allow different implementations of the hash function to be used without code changes, then I think it's a nice-to-have. Plus, this allows dependency injection.

ntallar replied:
Maybe I missunderstood something but doesn't the `PredefinedHasher` allow changing the hash function used?

javierdiaz72 replied:
Yes, it does, but to some extent. `PredefinedHasher` is parametrized by the name of a **predefined** algorithm (i.e. provided by `java.security.MessageDigest` such as MD5, SHA-1, etc.). If we want to write our own specific hasher (e.g. a mock hasher), then `Hasher` makes sense.

ntallar replied:
Are we expecting to write our own specific hasher in the near future? I agree that having a mock hasher could be benefitial, however I would prefer sticking to the simpler alternative if there's no need right now to make our code more complex.

ntallar replied:
Either way, maybe we could also move this `Hasher` classes and traits to a separate file, WDYT?

javierdiaz72 replied:
I already did it in PO-16 :) Please have a look at [this commit](https://github.com/atixlabs/praos_test/commit/32ab89754b56327a0b3a467781b88920db0c80f4).

ntallar commented on src/main/scala/io/iohk/praos/domain/Block.scala:
```diff
@@ -0,0 +1,21 @@
+  nonce       : (Seed, VrfProof),
+  signature   : Signature) {
+
+  def apply(genesisBlock: GenesisBlock): GenesisBlock = {
```
Maybe we should rename `GenesisBlock` to `VirtualGenesis` or something like that to differentiate it from the block 0? Also the block generated isn't necessarily a VirtualGenesis, right? There could be more blocks left on the epoch.

javierdiaz72 replied:
The Praos paper explicitly defines it as "genesis **block**" and differentiates it from a block in a block chain, whereas the Praos formalization calls it merely "genesis". I agree with you in that it's not properly a "block", but personally I don't like the adjective "virtual" (what would a "real genesis" be then?), so I would rather stick to the Praos formalization and call it "genesis".

You're right in that the result of applying a `Block` to a `GenesisBlock` is normally an "intermediate genesis", but I think it's not worth it to define another type for those intermediate results, which would turn out to have exactly the same structure.

javierdiaz72 replied:
Discussed with Nico, we will stick to the Praos paper.

ntallar replied:
Minor detail: Maybe we could add a comment on the `GenesisBlock` class regarding this uses of it?

javierdiaz72 replied:
Sounds good, will do.

javierdiaz72 replied:
Resolved.

ntallar commented on src/main/scala/io/iohk/praos/domain/Block.scala:
```diff
@@ -0,0 +1,21 @@
+case class Block(
+  state       : Option[Hasher#Digest],
+  slotNumber  : SlotNumber,
+  data        : List[Transaction],
```
Maybe we should rename this field to `transactions` or something similar?

javierdiaz72 replied:
Maybe. I stuck to the Praos paper, which seems to assume that a block may contain any data (since the protocol doesn't care about block contents), even though ultimately it will contain transactions. Following the same reasoning, If we want to rename `data` to `transactions`, then we should also rename `state` to `prevBlockHash` or similar.

javierdiaz72 replied:
Discussed with Nico, we will stick to the Praos paper.

ntallar commented on src/main/scala/io/iohk/praos/crypto/package.scala:
```diff
@@ -59,5 +59,44 @@ package object crypto {
+    * @note Currently implemented as the number obtained by concatenating the binary representation of both arguments.
+    */
+  def combineSeeds(x: Seed, y: Seed): Seed = {
+    Integer.parseInt(Integer.toBinaryString(x) ++ Integer.toBinaryString(y), 2)
```
Maybe this function should always return a positive number? We discussed with Mirko that it might be useful to do so as the random number AFAIK should always be positive (as in [here](https://github.com/atixlabs/praos_test/pull/6/files#diff-3f124fa27be1fc3a524517d6ca434401R15)). Do you know if maybe there is any place were that is not the case?

javierdiaz72 replied:
How could the concatenation of the binary representation of two positive numbers yield a negative number (apart from `Integer` overflow)? Please keep in mind that a `Seed` is a type alias for `RandomNumber` (which is assumed to be a positive number). Am I missing something?

ntallar replied:
You are right, I thought `Integer.parseInt` worked differently.
The case that I was thinking that could bring issues is for example `combineSeeds(Int.MaxValue, 1)` which seems to cause an exception, should we maybe address that case here?

javierdiaz72 replied:
As far as I understand, Mirko has already changed the implementation to `abs(x+y)`, correct?

ntallar commented on src/main/scala/io/iohk/praos/domain/Block.scala:
```diff
@@ -0,0 +1,21 @@
+
+
+case class Block(
+  state       : Option[Hasher#Digest],
```
Maybe we could define `state: Option[ByteString]`? The type of the hash function probably won't change, so we might avoid having `type Digest` in `Hasher`.

javierdiaz72 replied:
What if we want to use, say, `Array[Byte]` or similar in the future for whatever reason? IMHO, even when the result type of the hash function isn't likely to change, having a type `Hasher#Digest` is way more expressive than using a plain `ByteString`.

Commit:
Resolved PR comments.

ntallar commented on src/main/scala/io/iohk/praos/domain/RelativeStakeCalculator.scala:
```diff
@@ -29,9 +30,9 @@ object RelativeStakeCalculator {
-    * @note Precondition: stakeDistribution contains publicKey
     */
   def calculate(publicKey: Key, stakeDistribution: StakeDistribution): RelativeStake = {
+    require(stakeDistribution.contains(publicKey), "No such stakeholder in the stake distribution")
```
Maybe we should return that the stakeholder has no stake in that case?
I would prefer avoiding throwing an exception if possible, however this shouldn't be a blocker for merging this PR.

javierdiaz72 replied:
Hmm... IMHO we would be hiding a logic error if we return that the stakeholder has no stake in that case.

ntallar replied:
:thinking: I think that depends on the interpretation we give to this function.
We can leave it as is for now but we might probably reconsider it's interpretation when we start using it.

Commit:
Merge branch 'develop' of github.com:atixlabs/praos_test into feature/PO-6-Model-Block-Transaction-Blockchain

Commit:
Added further explanation about the use of the  GenesisBlock class.

For simplicity, GenesisBlocks are also used to store intermediate values when applying a sequence of blocks to a given GenesisBlock.

ntallar commented on src/main/scala/io/iohk/praos/domain/GenesisBlock.scala:
```diff
@@ -10,6 +10,12 @@ import io.iohk.praos.crypto.{Seed, combineSeeds}
 case class GenesisBlock(genesisDistribution: StakeDistribution, genesisNonce: Seed) {
+
+  /**
+    * @note The application of a block to a genesis block is *not* a genesis block when the applied block is not the
```
:+1:

ntallar reviewed (approved):
LGTM!
